### ISSUE TITLE
Untrue "missing return statement" after exhaustive match on enum

### ISSUE THREAD

**Issue by aleksanb** [python/mypy#16276]
**Bug Report**

Exhaustive match on enums fails if no intermediate alias of the match variable is given.

**To Reproduce**

```python
from enum import Enum

class Choices(str, Enum):
    A="A"
    B="B"
    
    
def matcher_implicit(a: str) -> str:
    match Choices(a):
        case Choices.A:
            return 'ok'
        case Choices.B:
            return 'schmok'
            

def matcher_alias(a: str) -> str:
    explicit_alias = Choices(a)
    match explicit_alias:
        case Choices.A:
            return 'ok'
        case Choices.B:
            return 'schmok'
```

You can run this on https://mypy-play.net/?mypy=latest&python=3.11, and see that only `matcher_implicit` fails, but there's no reason why it should!

**Expected Behavior**

Both should typecheck correctly 

**Actual Behavior**

The first one fails typechecking, with a fake "missing return statement" error emitted.

**Your Environment**

- Mypy version used: 1.5.1 and 1.6.0 (fails on both).
- sample fails on mypy-play.

**Comment by Sullynr2Wolf** [comment]
Hello! @aleksanb Can I try to fix the issue? I am currently doing my assignment for software engineering and am supposed to try solve an issue like this one. You can assign the issue to me and I will try my best to help.

**Comment by hauntsaninja** [comment]
Yes, don't ask, just open a PR.

(mypy is doing the equivalent of `if Choices(a) == Choices.A ... elif Choices(a) == Choices.B` where narrowing would be unsound because there's no guarantee `Choices` will return the same thing)

**Comment by hauntsaninja** [comment]
Duplicate of https://github.com/python/mypy/issues/12998

**Comment by BeefaloKing** [comment]
I don't think this is a duplicate of #12998 . I still have this issue with mypy 1.11+. Can we get this reopened?

**Comment by hauntsaninja** [comment]
This was indeed fixed in 1.10. Try it yourself in the playground. If you have some other issue, please open a new issue :-)

**Comment by BeefaloKing** [comment]
Ah, I see. The narrowing only works if you match against (e.g.) `Choices.A` rather than the literal `'A'`.
